sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamAggregationRel.java

```diff
@@ -111,6 +114,15 @@ public RelWriter explainTerms(RelWriter pw) {
     return pw;
   }
 
+  @Override
+  public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery metadata) {
+    RelNode child = getInput();
+    Double rowCnt = metadata.getRowCount(child);
```

Comment:
Because there is no support to get the relatively estimate row count yet, I suggest we don't work on CBO but focus on something we can control (e.g. logical optimization). For example, we can work on figuring out whether rules work in  [BeamRuleSets.java](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/planner/BeamRuleSets.java). I have seen an issue that we add a rule int the list but that rule triggers a bug at a moment. Also some rules definitely are not working because of traits setup.

Reply:
A `BoundedSource` does have size estimation that we might be able to use. What I think is important is the ability to correctly guide Calcite to apply the desired rules.